Env.Extend() result shares state

**Describe the bug**
This is a minor issue that prevents using a shared parent Env to build multiple child Envs.

Env.Extend() returns a new Env that shares the underlying slices of the original Env. If I extend the same Env more than once then previous extended Envs can be modified by the current one. This is because they append into the same slice.

**To Reproduce**
Check which components this affects:

- [ ]  parser
- [ ]  checker
- [ ]  interpreter

Test setup:
```go
func TestEnvExtends(t *testing.T) {
	baseEnv, err := cel.NewEnv(cel.Declarations(
		decls.NewIdent("age", decls.Int, nil),
		decls.NewIdent("gender", decls.String, nil),
		decls.NewIdent("country", decls.String, nil)))
	if err != nil {
		t.Fatal(err)
	}

	env1, err := baseEnv.Extend(cel.Declarations(
		decls.NewIdent("name", decls.String, nil)))
	if err != nil {
		t.Fatal(err)
	}

	env2, err := baseEnv.Extend(cel.Declarations(
		decls.NewIdent("group", decls.String, nil)))
	if err != nil {
		t.Fatal(err)
	}

	if _, issues := env2.Compile("size(group) > 10"); issues.Err() != nil {
		t.Fatal(issues.Err())
	}

	// I expect this to work but env1 now has "group" instead of "name"
	if _, issues := env1.Compile("size(name) > 10"); issues.Err() != nil {
		t.Fatal(issues.Err())
	}
}
```
```
--- FAIL: TestEnvExtends (0.01s)
    cel_test.go:37: ERROR: <input>:1:6: undeclared reference to 'name' (in container '')
         | size(name) > 10
         | .....^
```

**Expected behavior**
In the test case I expect env1 to have the declarations of baseEnv + name and env2 to have the declarations of baseEnv + group.

I'd also expect similar behavior for other configurable slice fields.

**Additional context**
I'm reporting this as an Env.Extend() issue but it could also be solved if the EnvOptions made new copies of slices instead of appending.

--- a/cel/BUILD.bazel
+++ b/cel/BUILD.bazel
@@ -52,6 +52,8 @@ go_test(
         "//common/types/ref:go_default_library",
         "//common/types/traits:go_default_library",
         "//interpreter/functions:go_default_library",
+        "//test/proto2pb:go_default_library",
+        "//test/proto3pb:go_default_library",
         "@io_bazel_rules_go//proto/wkt:descriptor_go_proto",
         "@org_golang_google_genproto//googleapis/api/expr/v1alpha1:go_default_library",
     ],

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -705,9 +705,25 @@ func Test_EnvExtension(t *testing.T) {
 				decls.NewObjectType("google.api.expr.v1alpha1.Expr"), nil),
 		),
 	)
-	e2, _ := e.Extend()
+	e2, _ := e.Extend(
+		CustomTypeAdapter(types.DefaultTypeAdapter),
+		Types(&proto3pb.TestAllTypes{}),
+	)
 	if e == e2 {
 		t.Error("got object equality, wanted separate objects")
+	}
+	if e.TypeAdapter() == e2.TypeAdapter() {
+		t.Error("got the same type adapter, wanted isolated instances.")
+	}
+	if e.TypeProvider() == e2.TypeProvider() {
+		t.Error("got the same type provider, wanted isolated instances.")
+	}
+	e3, _ := e2.Extend()
+	if e2.TypeAdapter() != e3.TypeAdapter() {
+		t.Error("got different type adapters, wanted immutable adapter reference")
+	}
+	if e2.TypeProvider() == e3.TypeProvider() {
+		t.Error("got the same type provider, wanted isolated instances.")
 	}
 }
 

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -730,9 +730,14 @@ func Test_EnvExtensionIsolation(t *testing.T) {
 	if _, issues := env2.Compile("size(group) > 10"); issues.Err() != nil {
 		t.Fatal(issues.Err())
 	}
-	// I expect this to work but env1 now has "group" instead of "name"
+	if _, issues := env2.Compile("size(name) > 10"); issues.Err() == nil {
+		t.Fatal("env2 contains 'name', but should not")
+	}
 	if _, issues := env1.Compile("size(name) > 10"); issues.Err() != nil {
 		t.Fatal(issues.Err())
+	}
+	if _, issues := env1.Compile("size(group) > 10"); issues.Err() == nil {
+		t.Fatal("env1 contains 'group', but should not")
 	}
 }
 

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -709,6 +709,33 @@ func Test_EnvExtension(t *testing.T) {
 	}
 }
 
+func Test_EnvExtensionIsolation(t *testing.T) {
+	baseEnv, err := NewEnv(Declarations(
+		decls.NewIdent("age", decls.Int, nil),
+		decls.NewIdent("gender", decls.String, nil),
+		decls.NewIdent("country", decls.String, nil)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	env1, err := baseEnv.Extend(Declarations(
+		decls.NewIdent("name", decls.String, nil)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	env2, err := baseEnv.Extend(Declarations(
+		decls.NewIdent("group", decls.String, nil)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, issues := env2.Compile("size(group) > 10"); issues.Err() != nil {
+		t.Fatal(issues.Err())
+	}
+	// I expect this to work but env1 now has "group" instead of "name"
+	if _, issues := env1.Compile("size(name) > 10"); issues.Err() != nil {
+		t.Fatal(issues.Err())
+	}
+}
+
 func Test_ParseAndCheckConcurrently(t *testing.T) {
 	e, _ := NewEnv(
 		Container("google.api.expr.v1alpha1"),

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -33,6 +33,8 @@ import (
 	"github.com/google/cel-go/parser"
 
 	descpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	proto2pb "github.com/google/cel-go/test/proto2pb"
+	proto3pb "github.com/google/cel-go/test/proto3pb"
 	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 )
 
@@ -710,34 +712,55 @@ func Test_EnvExtension(t *testing.T) {
 }
 
 func Test_EnvExtensionIsolation(t *testing.T) {
-	baseEnv, err := NewEnv(Declarations(
-		decls.NewIdent("age", decls.Int, nil),
-		decls.NewIdent("gender", decls.String, nil),
-		decls.NewIdent("country", decls.String, nil)))
+	baseEnv, err := NewEnv(
+		Container("google.expr"),
+		Declarations(
+			decls.NewIdent("age", decls.Int, nil),
+			decls.NewIdent("gender", decls.String, nil),
+			decls.NewIdent("country", decls.String, nil),
+		),
+	)
 	if err != nil {
 		t.Fatal(err)
 	}
-	env1, err := baseEnv.Extend(Declarations(
-		decls.NewIdent("name", decls.String, nil)))
+	env1, err := baseEnv.Extend(
+		Types(&proto2pb.TestAllTypes{}),
+		Declarations(decls.NewIdent("name", decls.String, nil)))
 	if err != nil {
 		t.Fatal(err)
 	}
-	env2, err := baseEnv.Extend(Declarations(
-		decls.NewIdent("group", decls.String, nil)))
+	env2, err := baseEnv.Extend(
+		Types(&proto3pb.TestAllTypes{}),
+		Declarations(decls.NewIdent("group", decls.String, nil)))
 	if err != nil {
 		t.Fatal(err)
 	}
-	if _, issues := env2.Compile("size(group) > 10"); issues.Err() != nil {
+	_, issues := env2.Compile(`size(group) > 10
+		&& !has(proto3.test.TestAllTypes{}.single_int32)`)
+	if issues.Err() != nil {
 		t.Fatal(issues.Err())
 	}
-	if _, issues := env2.Compile("size(name) > 10"); issues.Err() == nil {
+	_, issues = env2.Compile(`size(name) > 10`)
+	if issues.Err() == nil {
 		t.Fatal("env2 contains 'name', but should not")
 	}
-	if _, issues := env1.Compile("size(name) > 10"); issues.Err() != nil {
+	_, issues = env2.Compile(`!has(proto2.test.TestAllTypes{}.single_int32)`)
+	if issues.Err() == nil {
+		t.Fatal("env2 contains 'proto2.test.TestAllTypes', but should not")
+	}
+
+	_, issues = env1.Compile(`size(name) > 10
+		&& !has(proto2.test.TestAllTypes{}.single_int32)`)
+	if issues.Err() != nil {
 		t.Fatal(issues.Err())
 	}
-	if _, issues := env1.Compile("size(group) > 10"); issues.Err() == nil {
+	_, issues = env1.Compile("size(group) > 10")
+	if issues.Err() == nil {
 		t.Fatal("env1 contains 'group', but should not")
+	}
+	_, issues = env1.Compile(`!has(proto3.test.TestAllTypes{}.single_int32)`)
+	if issues.Err() == nil {
+		t.Fatal("env1 contains 'proto3.test.TestAllTypes', but should not")
 	}
 }
 

--- a/cel/env.go
+++ b/cel/env.go
@@ -209,6 +209,7 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	if e.chkErr != nil {
 		return nil, e.chkErr
 	}
+	// Copy slices.
 	decsCopy := make([]*exprpb.Decl, len(e.declarations))
 	macsCopy := make([]parser.Macro, len(e.macros))
 	progOptsCopy := make([]ProgramOption, len(e.progOpts))
@@ -216,14 +217,29 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	copy(macsCopy, e.macros)
 	copy(progOptsCopy, e.progOpts)
 
+	// Copy the adapter / provider if they appear to be mutable.
+	var adapter ref.TypeAdapter
+	var provider ref.TypeProvider
+	adapterReg, isAdapterReg := e.adapter.(ref.TypeRegistry)
+	providerReg, isProviderReg := e.provider.(ref.TypeRegistry)
+	if isAdapterReg && isProviderReg && adapterReg == providerReg {
+		reg := providerReg.Copy()
+		adapter = reg
+		provider = reg
+	} else if isProviderReg {
+		provider = providerReg.Copy()
+	} else if isAdapterReg {
+		adapter = adapterReg.Copy()
+	}
+
 	ext := &Env{
 		declarations:                   decsCopy,
 		macros:                         macsCopy,
 		progOpts:                       progOptsCopy,
-		adapter:                        e.adapter,
+		adapter:                        adapter,
 		enableDynamicAggregateLiterals: e.enableDynamicAggregateLiterals,
 		pkg:                            e.pkg,
-		provider:                       e.provider,
+		provider:                       provider,
 	}
 	return ext.configure(opts)
 }

--- a/cel/env.go
+++ b/cel/env.go
@@ -205,6 +205,12 @@ func (e *Env) CompileSource(src common.Source) (*Ast, *Issues) {
 }
 
 // Extend the current environment with additional options to produce a new Env.
+//
+// Note, the extended Env value should not share memory with the original. It is possible, however,
+// that a CustomTypeAdapter or CustomTypeProvider options could provide values which are mutable.
+// To ensure separation of state between extended environments either make sure the TypeAdapter and
+// TypeProvider are immutable, or that their underlying implementations are based on the
+// ref.TypeRegistry which provides a Copy method which will be invoked by this method.
 func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	if e.chkErr != nil {
 		return nil, e.chkErr
@@ -222,6 +228,11 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	var provider ref.TypeProvider
 	adapterReg, isAdapterReg := e.adapter.(ref.TypeRegistry)
 	providerReg, isProviderReg := e.provider.(ref.TypeRegistry)
+	// In most cases the provider and adapter will be a ref.TypeRegistry;
+	// however, in the rare cases where they are not, they are assumed to
+	// be immutable. Since it is possible to set the TypeProvider separately
+	// from the TypeAdapter, the possible configurations which could use a
+	// TypeRegistry as the base implementation are captured below.
 	if isAdapterReg && isProviderReg && adapterReg == providerReg {
 		reg := providerReg.Copy()
 		adapter = reg

--- a/cel/env.go
+++ b/cel/env.go
@@ -209,13 +209,20 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	if e.chkErr != nil {
 		return nil, e.chkErr
 	}
+	decsCopy := make([]*exprpb.Decl, len(e.declarations))
+	macsCopy := make([]parser.Macro, len(e.macros))
+	progOptsCopy := make([]ProgramOption, len(e.progOpts))
+	copy(decsCopy, e.declarations)
+	copy(macsCopy, e.macros)
+	copy(progOptsCopy, e.progOpts)
+
 	ext := &Env{
-		declarations:                   e.declarations,
+		declarations:                   decsCopy,
+		macros:                         macsCopy,
+		progOpts:                       progOptsCopy,
 		adapter:                        e.adapter,
 		enableDynamicAggregateLiterals: e.enableDynamicAggregateLiterals,
-		macros:                         e.macros,
 		pkg:                            e.pkg,
-		progOpts:                       e.progOpts,
 		provider:                       e.provider,
 	}
 	return ext.configure(opts)

--- a/cel/env.go
+++ b/cel/env.go
@@ -224,8 +224,8 @@ func (e *Env) Extend(opts ...EnvOption) (*Env, error) {
 	copy(progOptsCopy, e.progOpts)
 
 	// Copy the adapter / provider if they appear to be mutable.
-	var adapter ref.TypeAdapter
-	var provider ref.TypeProvider
+	adapter := e.adapter
+	provider := e.provider
 	adapterReg, isAdapterReg := e.adapter.(ref.TypeRegistry)
 	providerReg, isProviderReg := e.provider.(ref.TypeRegistry)
 	// In most cases the provider and adapter will be a ref.TypeRegistry;

--- a/common/types/pb/pb.go
+++ b/common/types/pb/pb.go
@@ -61,6 +61,15 @@ func NewDb() *Db {
 	return pbdb
 }
 
+// Copy creates a copy of the current database with its own internal descriptor mapping.
+func (pbdb *Db) Copy() *Db {
+	copy := NewDb()
+	for k, v := range pbdb.revFileDescriptorMap {
+		copy.revFileDescriptorMap[k] = v
+	}
+	return copy
+}
+
 // RegisterDescriptor produces a `FileDescription` from a `FileDescriptorProto` and registers the
 // message and enum types into the `pb.Db`.
 func (pbdb *Db) RegisterDescriptor(fileDesc *descpb.FileDescriptorProto) (*FileDescription, error) {

--- a/common/types/provider.go
+++ b/common/types/provider.go
@@ -77,6 +77,19 @@ func NewEmptyRegistry() ref.TypeRegistry {
 	}
 }
 
+// Copy implements the ref.TypeRegistry interface method which copies the current state of the
+// registry into its own memory space.
+func (p *protoTypeRegistry) Copy() ref.TypeRegistry {
+	copy := &protoTypeRegistry{
+		revTypeMap: make(map[string]ref.Type),
+		pbdb:       p.pbdb.Copy(),
+	}
+	for k, v := range p.revTypeMap {
+		copy.revTypeMap[k] = v
+	}
+	return copy
+}
+
 func (p *protoTypeRegistry) EnumValue(enumName string) ref.Val {
 	enumVal, err := p.pbdb.DescribeEnum(enumName)
 	if err != nil {

--- a/common/types/ref/provider.go
+++ b/common/types/ref/provider.go
@@ -78,6 +78,9 @@ type TypeRegistry interface {
 	// If a type is provided more than once with an alternative definition, the
 	// call will result in an error.
 	RegisterType(types ...Type) error
+
+	// Copy the TypeRegistry and return a new registry whose mutable state is isolated.
+	Copy() TypeRegistry
 }
 
 // FieldType represents a field's type value and whether that field supports

--- a/test/proto2pb/BUILD.bazel
+++ b/test/proto2pb/BUILD.bazel
@@ -3,6 +3,7 @@ load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 
 package(
     default_visibility = [
+        "//cel:__subpackages__",
         "//checker:__subpackages__",
         "//common:__subpackages__",
         "//interpreter:__subpackages__",

--- a/test/proto3pb/BUILD.bazel
+++ b/test/proto3pb/BUILD.bazel
@@ -3,6 +3,7 @@ load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 
 package(
     default_visibility = [
+        "//cel:__subpackages__",
         "//checker:__subpackages__",
         "//common:__subpackages__",
         "//interpreter:__subpackages__",

